sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,150 @@
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;
+
+    // Let's be sure to end any running trace
+    if (activeTransaction != null) onTransactionEnd(activeTransaction);
+
+    traceFile = FileUtils.resolve(traceFilesDir, "sentry-" + UUID.randomUUID() + ".trace");
+
+    if (traceFile == null) {
+      options.getLogger().log(SentryLevel.DEBUG, "Could not create a trace file");
+      return;
+    }
+
+    if (traceFile.exists()) {
+      options
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Trace file already exists: %s", traceFile.getPath());
+      return;
+    }
+
+    long intervalMs = options.getProfilingTracesIntervalMillis();
+    if (intervalMs <= 0) {
+      options
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Profiling trace interval is set to %d milliseconds", intervalMs);
+      return;
+    }
+
+    // We stop the trace after 30 seconds, since such a long trace is very probably a trace
+    // that will never end due to an error
+    SentryExecutors.tracingExecutor.schedule(
+        () -> onTransactionEnd(transaction), 30_000, MILLISECONDS);
+
+    startedMethodTracing = true;
+    int intervalUs = (int) MILLISECONDS.toMicros(intervalMs);
+    activeTransaction = transaction;
+    Debug.startMethodTracingSampling(traceFile.getPath(), BUFFER_SIZE_BYTES, intervalUs);
+  }
+
+  @Override
+  public synchronized void onTransactionEnd(ITransaction transaction) {
+    // In case a previous timeout tries to end a newer transaction we simply ignore it
+    if (transaction != activeTransaction) return;
+
+    if (startedMethodTracing) {
+      startedMethodTracing = false;
+      activeTransaction = null;
+
+      Debug.stopMethodTracing();
+
+      if (traceFile == null || !traceFile.exists()) {
+        options
+            .getLogger()
+            .log(
+                SentryLevel.DEBUG,
+                "Trace file %s does not exists",
+                traceFile == null ? "null" : traceFile.getPath());
+        return;
+      }
+
+      // todo should I use transaction.getEventId() instead of new SentryId()?
+      //  Or should I add the transaction id as an header to the envelope?
+      //  Or should I simply ignore the transaction entirely (wouldn't make any sense)?
+      //  And how to check if a trace is from a startup?
+      try {
+        SentryEnvelope envelope =
+            SentryEnvelope.from(
+                new SentryId(),
+                traceFile.getPath(),
+                traceFile.getName(),
+                options.getSdkVersion(),
+                options.getMaxTraceFileSize(),
+                true);
+        hub.captureEnvelope(envelope);
+      } catch (IOException e) {
+        options.getLogger().log(SentryLevel.ERROR, "Failed to capture the trace.", e);
+        // We don't return out of this function here, so we can go on and clean the variables
+      }
+    }
+
+    if (traceFile != null) traceFile.delete();
```

Comment:
This is likely deleting the file even before the SDK got the chance to send it over to the network since the EnvelopeItem is read async.